### PR TITLE
Operator utils module

### PR DESCRIPTION
**Pull request by chmeliik**
* OSBS-8636

TODO:
- [x] Some TODOs in code and specfile

Maintainers will complete the following section:
- [x] Commit messages are descriptive enough
- [x] "Signed-off-by:" line is present in each commit
- [x] Code coverage from testing does not decrease and new code is covered
- [x] JSON/YAML configuration changes are updated in the relevant schema
- [x] Changes to metadata also update the documentation for the metadata
- [x] Pull request includes link to an osbs-docs PR for user documentation updates
- [x] New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/operator_util.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        Make an OperatorCSV from a file
+
+        :param path: Path to file
+        :return: OperatorCSV
+        """
+        with open(path) as f:
+            data = yaml.load(f)
+            return cls(path, data)
+
+    def write(self):
```
Comment (MartinBasti) [inline atomic_reactor/operator_util.py]: maybe `dump` may express it better

File: atomic_reactor/operator_util.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+            for f in files:
+                if f.endswith(".yaml") or f.endswith(".yml"):
+                    yield os.path.join(d, f)
+
+    @classmethod
+    def _get_csvs(cls, yaml_files):
+        for f in yaml_files:
+            try:
+                yield OperatorCSV.from_file(f)
+            except ValueError:
```
Comment (MartinBasti) [inline atomic_reactor/operator_util.py]: Can we raise specific error, like `NotOperatorCSVFile`

File: atomic_reactor/operator_util.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        self._replace_env_pullspecs(replacement_pullspecs)
+        self._replace_container_image_pullspecs(replacement_pullspecs)
+
+    def replace_pullspecs_everywhere(self, replacement_pullspecs):
+        """
+        Replace all pullspecs found anywhere in data
+
+        :param replacement_pullspecs: mapping of pullspec -> replacment
+        """
+        log_template = "{path} - Replaced pullspec: {old} -> {new}"
```
Comment (MartinBasti) [inline atomic_reactor/operator_util.py]: Why? Can we have just constant `LOG_TEMPLATE = "%s - Replaced pullspec: %s --> %s"` without interpolation adn conditions?

File: atomic_reactor/operator_util.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            for e in c.get("env", []) if e["name"].startswith("RELATED_IMAGE_")
+        ]
+        for env in envs:
+            if "valueFrom" in env:
+                msg = '{}: "valueFrom" references are not supported'.format(env["name"])
+                raise RuntimeError(msg)
+        return envs
+
+    def _get_pullspec(self, obj, key, log_template=None):
+        # TODO: better name than obj?
```
Comment (MartinBasti) [inline atomic_reactor/operator_util.py]: fine by me

File: atomic_reactor/operator_util.py (hunk 3 above)
Comment (chmeliik) [reply atomic_reactor/operator_util.py]: I thought it would be nice to log where exactly a pullspec was found/replaced, but I did manage to get rid of string interpolation FWIW

File: atomic_reactor/operator_util.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        deployments_path = ("spec", "install", "spec", "deployments")
+        deployments = chain_get(self.data, deployments_path, default=[])
+
+        containers_path = ("spec", "template", "spec", "containers")
+        containers = [
+            c for d in deployments for c in chain_get(d, containers_path, default=[])
+        ]
+        return containers
+
+    def _get_envs(self):
```
Comment (MartinBasti) [inline atomic_reactor/operator_util.py]: maybe `get_related_image_envs` ?

**Comment by chmeliik** [comment]
Fixed specfile (ruamel package on RHEL)

File: atomic_reactor/operator_util.py (hunk 5 above)
Comment (chmeliik) [reply atomic_reactor/operator_util.py]: Renamed, also realized test for "valueFrom" in env definition was missing, added

**Comment by chmeliik** [comment]
Removed resolved TODO

**Comment by chmeliik** [comment]
Added more details in commit messages

File: atomic_reactor/operator_util.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+        return containers
+
+    def _get_related_image_envs(self):
+        containers = self._get_containers()
+        envs = [
+            e for c in containers
+            for e in c.get("env", []) if e["name"].startswith("RELATED_IMAGE_")
+        ]
+        for env in envs:
+            if "valueFrom" in env:
```
Comment (rcerven) [inline atomic_reactor/operator_util.py]: I think we should just ignore those, but not failing

@MartinBasti  wdyt ?

Comment (MartinBasti) [reply atomic_reactor/operator_util.py]: Hmm, yup we should just ignore it. Good catch

File: atomic_reactor/util.py [hunk 7]
```diff
@@ -1611,3 +1611,12 @@ def get_platform_config(platform, build_annotations):
     cm_key = cm_key_tmp[cmlen:]
 
     return cm_key, cm_frag_key
+
+
+def chain_get(d, keys, default=None):
+    for k in keys:
+        try:
+            d = d[k]
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply atomic_reactor/util.py]: @ben-alkov; This function is shamelessly stolen from osbs-client (with minor changes). I think you have the right idea about what it does, but I'll try to give an example just to be sure:

Consider the following yaml file:
```yaml
spec:
  relatedImages:
    - name: foo
      image: registry/foo:1
```
Using `jq`, you would query for the image pullspec like this (after converting from yaml to json)
```bash
$ cat file.json | jq '.spec.relatedImages[0].image'
"registry/foo:1"
```
With this function, you can achieve the same thing for a python dict:
```python
>>> chain_get(data, ["spec", "relatedImages", 0, "image"])
"registry/foo:1"

# basically the same as
data["spec"]["relatedImages"][0]["image"]
# but returns None if any of the lookups would fail
```

> Also, single letter variable names :rage:

`d` as a variable name for "this is a dictionary with futher unspecified content" is pretty common, but it's true that the meaning may change inside the loop (sometimes, it may be a `list`, at the end, it will usually be a `str`, `int`, `float` or `bool`). I could rename to something that better reflects that, but best I can think of is `obj` for object

File: atomic_reactor/operator_util.py (hunk 6 above)
Comment (chmeliik) [reply atomic_reactor/operator_util.py]: Should we? It is still a `RELATED_IMAGE_*` environment variable with some pullspec, we just do not know what that pullspec is. Is it ok for that to be in the final operator manifest?

Alternatively, I could try to implement the `valueFrom` logic, I haven't looked at the specification (if there is any) but it shouldn't be impossibly difficult.

**Comment by chmeliik** [comment]
Added a docstring to `chain_get` and improved variable naming

File: atomic_reactor/operator_util.py (hunk 6 above)
Comment (MartinBasti) [reply atomic_reactor/operator_util.py]: Well, it's reference we cannot catch everything. But on other hand we can set this as requirement to not use references for that in OSBS and play more safe.

File: atomic_reactor/util.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ben-alkov) [reply atomic_reactor/util.py]: @chmeliik; Very clear, thanks!
